### ISSUE TITLE
Does not compile on raspbian

### ISSUE THREAD

**Issue by raphaelm22** [Azure/azure-storage-fuse#14]
I can not compile on raspbian
```
./build.sh: line 9: cmake3: comando não encontrado
-- The C compiler identification is GNU 4.9.2
-- The CXX compiler identification is GNU 4.9.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CURL: /usr/lib/arm-linux-gnueabihf/libcurl.so (found version "7.38.0")
-- Found GnuTLS: /usr/lib/arm-linux-gnueabihf/libgnutls.so (found version "3.3.8")
-- Configuring done
-- Generating done
-- Build files have been written to: /root/azure-storage-fuse/azure-storage-fuse/build
Scanning dependencies of target blobfuse
[  3%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/blobfuse.cpp.o
[  6%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/fileapis.cpp.o
/root/azure-storage-fuse/azure-storage-fuse/blobfuse/fileapis.cpp: In function ‘int azs_truncate(const char*, off_t)’:
/root/azure-storage-fuse/azure-storage-fuse/blobfuse/fileapis.cpp:520:85: error: format ‘%ld’ expects argument of type ‘long int’, but argument 4 has type ‘off_t {aka long long int}’ [-Werror=format=]
         fprintf(stdout, "azs_truncate called.  Path = %s, offset = %ld\n", path, off);
                                                                                     ^
cc1plus: all warnings being treated as errors
CMakeFiles/blobfuse.dir/build.make:86: recipe for target 'CMakeFiles/blobfuse.dir/blobfuse/fileapis.cpp.o' failed
make[2]: *** [CMakeFiles/blobfuse.dir/blobfuse/fileapis.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/blobfuse.dir/all' failed
make[1]: *** [CMakeFiles/blobfuse.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2

```

**Comment by seguler** [comment]
What is your gcc and cmake version ?

**Comment by AlexanderYukhanov** [comment]
it's because raspberry pi is 32 bits, on 32 bits long int is 32 bits and long long is 64 bits.

**Comment by seguler** [comment]
Oh wow - what was your scenario @raphaelm22 to run this on a raspberry ?

**Comment by raphaelm22** [comment]
Hi @seguler, follows the versions.

```
raphael@raspberrypi:~ $ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/arm-linux-gnueabihf/4.9/lto-wrapper
Target: arm-linux-gnueabihf
Configured with: ../src/configure -v --with-pkgversion='Raspbian 4.9.2-10' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libitm --disable-libquadmath --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-armhf/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-armhf --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-armhf --with-arch-directory=arm --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-sjlj-exceptions --with-arch=armv6 --with-fpu=vfp --with-float=hard --enable-checking=release --build=arm-linux-gnueabihf --host=arm-linux-gnueabihf --target=arm-linux-gnueabihf
Thread model: posix
gcc version 4.9.2 (Raspbian 4.9.2-10)
```

```
raphael@raspberrypi:~ $ cmake -version
cmake version 3.6.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

I use raspbian as my central media and share personal files in my house, such as photos and family videos :).

The idea is to map the storage and make samba available so that anyone can back up easily on laptops with windows.

Another idea is to create backup routines for external hard drives connected to raspbean, I know I could do this with AzCopy or azurecopy, but I really liked the idea of azure-storage-fuse.

**Comment by polarapfel** [comment]
@AlexanderYukhanov is right.

Check this for some more context: https://stackoverflow.com/a/14351239/1341177

So, either change the type in that function from off_t to off64_t or compile with _FILE_OFFSET_BITS == 64. That's worth a try.

Hope that helps.

**Comment by raphaelm22** [comment]
Hi. I tried editing the build.sh file by changing line 11 to

`cmake -D_FILE_OFFSET_BITS=64-DCMAKE_BUILD_TYPE=Release ..`

So I received this warning at output:

```
...
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    _FILE_OFFSET_BITS


-- Build files have been written to: /root/azure-storage-fuse/build
Scanning dependencies of target blobfuse
[  3%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/blobfuse.cpp.o
[  6%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/fileapis.cpp.o
/root/azure-storage-fuse/blobfuse/fileapis.cpp: In function ‘int azs_truncate(const char*, off_t)’:
/root/azure-storage-fuse/blobfuse/fileapis.cpp:520:85: error: format ‘%ld’ expects argument of type ‘long int’, but argument 4 has type ‘off_t {aka long long int}’ [-Werror=format=]
         fprintf(stdout, "azs_truncate called.  Path = %s, offset = %ld\n", path, off);
                                                                                     ^
cc1plus: all warnings being treated as errors
CMakeFiles/blobfuse.dir/build.make:86: recipe for target 'CMakeFiles/blobfuse.dir/blobfuse/fileapis.cpp.o' failed
make[2]: *** [CMakeFiles/blobfuse.dir/blobfuse/fileapis.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/blobfuse.dir/all' failed
make[1]: *** [CMakeFiles/blobfuse.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2

```

**Comment by polarapfel** [comment]
Mmh... It didn't take the compiler flag. I wonder if the flag needs to go someplace else. I am not too familiar with how cmake works in particular. I'll take a look.

**Comment by AlexanderYukhanov** [comment]
Why you don't want just to fix the code instead? Let's just replace all those fprintfs with iostream or boost::format?

**Comment by AlexanderYukhanov** [comment]
Submitted a pull request https://github.com/Azure/azure-storage-fuse/pull/19 to resolve 32 bits compilation problem.

**Comment by raphaelm22** [comment]
Hi @AlexanderYukhanov 
I tried compiling on raspberry

```
root@raspberrypi:~/azure-storage-fuse # git pull origin pull/19/head
remote: Counting objects: 175, done.
remote: Compressing objects: 100% (35/35), done.
remote: Total 175 (delta 23), reused 38 (delta 17), pack-reused 123
Receiving objects: 100% (175/175), 152.79 KiB | 0 bytes/s, done.
Resolving deltas: 100% (68/68), done.
From https://github.com/Azure/azure-storage-fuse
 * branch            refs/pull/19/head -> FETCH_HEAD
root@raspberrypi:~/azure-storage-fuse # ./build.sh
./build.sh: line 9: cmake3: comando não encontrado
-- The C compiler identification is GNU 4.9.2
-- The CXX compiler identification is GNU 4.9.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CURL: /usr/lib/arm-linux-gnueabihf/libcurl.so (found version "7.38.0")
-- Found GnuTLS: /usr/lib/arm-linux-gnueabihf/libgnutls.so (found version "3.3.8")
-- Configuring done
-- Generating done
-- Build files have been written to: /root/azure-storage-fuse/build
Scanning dependencies of target blobfuse
[  3%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/blobfuse.cpp.o
[  6%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/fileapis.cpp.o
[  9%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/directoryapis.cpp.o
[ 12%] Building CXX object CMakeFiles/blobfuse.dir/blobfuse/utilities.cpp.o
[ 15%] Building CXX object CMakeFiles/blobfuse.dir/azure-storage-cpp-lite/src/base64.cpp.o
/root/azure-storage-fuse/azure-storage-cpp-lite/src/base64.cpp: In function ‘std::vector<unsigned char> microsoft_azure::storage::from_base64(const string&)’:
/root/azure-storage-fuse/azure-storage-cpp-lite/src/base64.cpp:100:28: error: comparison is always false due to limited range of data type [-Werror=type-limits]
                     if (ch < 0)
                            ^
/root/azure-storage-fuse/azure-storage-cpp-lite/src/base64.cpp:120:37: error: comparison is always false due to limited range of data type [-Werror=type-limits]
                             if (ch2 < 0)
                                     ^
cc1plus: all warnings being treated as errors
CMakeFiles/blobfuse.dir/build.make:158: recipe for target 'CMakeFiles/blobfuse.dir/azure-storage-cpp-lite/src/base64.cpp.o' failed
make[2]: *** [CMakeFiles/blobfuse.dir/azure-storage-cpp-lite/src/base64.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/blobfuse.dir/all' failed
make[1]: *** [CMakeFiles/blobfuse.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by AlexanderYukhanov** [comment]
Yep, it's another cross platform issue. On arm 32 char is unsigned, but on x86 it's signed. There are two options: 1) we can provide -fsigned-char flag to make it signed or 2) cast ch to signed char.
Will reanimate my raspberry pi and fix this issue as well by Monday

**Comment by AlexanderYukhanov** [comment]
Added a new commit with the fix. Compiles and works on raspberry pi.

**Comment by raphaelm22** [comment]
Thanks @AlexanderYukhanov, it is working perfectly.

**Comment by seguler** [comment]
Fix is now in dev branch with PR #28. This will be released end of December as a binary.
